dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DefaultDeduplicationService.java

```diff
@@ -101,6 +105,70 @@ public void updatePotentialDuplicate( PotentialDuplicate potentialDuplicate )
         potentialDuplicateStore.update( potentialDuplicate );
     }
 
+    @Override
+    public boolean isAutoMergeable( PotentialDuplicate potentialDuplicate )
+    {
+        TrackedEntityInstance trackedEntityInstanceA = Optional.ofNullable( trackedEntityInstanceService
+            .getTrackedEntityInstance( potentialDuplicate.getTeiA() ) )
+            .orElseThrow( () -> new PotentialDuplicateException(
+                "No tracked entity instance found with id '" + potentialDuplicate.getTeiA() + "'." ) );
+
+        TrackedEntityInstance trackedEntityInstanceB = Optional.ofNullable( trackedEntityInstanceService
+            .getTrackedEntityInstance( potentialDuplicate.getTeiB() ) )
+            .orElseThrow( () -> new PotentialDuplicateException(
+                "No tracked entity instance found with id '" + potentialDuplicate.getTeiB() + "'." ) );
+
+        if ( !trackedEntityInstanceA.getTrackedEntityType().equals( trackedEntityInstanceB.getTrackedEntityType() ) )
+        {
+            return false;
+        }
+
+        if ( trackedEntityInstanceA.isDeleted() || trackedEntityInstanceB.isDeleted() )
+        {
+            return false;
+        }
+
+        if ( trackedEntityInstanceA.getRelationshipItems().stream()
```

Comment:
A couple of notes:

1. If a relationship between A and B is found, it is still mergable. But that relationship will be removed, since we cant have self-referencing relationships.

2. trackedEntityInstanceA.getRelationshipItems() returns any item of that tei. an item can be either part of the TO or FROM of a relationship. In this case, teiA might be the "to" part of the relationship, so you are just comparing to itself. What you should be doing here to comparing both to and from.

In either case, I don't think we need this specific validation, since we will just remove it if this is true